src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java

```
@@ -0,0 +1,102 @@
+    return ctpClient;
+  }
+
+  protected CompletableFuture<Void> fetchAndFillReferenceIdToKeyCache(
+      @Nonnull final Set<String> ids, @Nonnull final GraphQlQueryResources requestType) {
+
+    final Set<String> nonCachedReferenceIds = getNonCachedReferenceIds(ids);
+    if (nonCachedReferenceIds.isEmpty()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    final List<List<String>> chunkedIds = ChunkUtils.chunk(nonCachedReferenceIds, CHUNK_SIZE);
+
+    List<ResourceIdsGraphQlRequest> collectedRequests =
+        createResourceIdsGraphQlRequests(chunkedIds, requestType);
+
+    return ChunkUtils.executeChunks(getCtpClient(), collectedRequests)
+        .thenApply(ChunkUtils::flattenGraphQLBaseResults)
+        .thenCompose(
+            results -> {
+              try {
+                cacheResourceReferenceKeys(results);
+              } catch (ReferenceTransformException referenceTransformException) {
+                return exceptionallyCompletedFuture(referenceTransformException);
+              }
+              return CompletableFuture.completedFuture(null);
+            });
+  }
+
+  @Nonnull
+  protected List<ResourceIdsGraphQlRequest> createResourceIdsGraphQlRequests(
+      @Nonnull final List<List<String>> chunkedIds,
+      @Nonnull final GraphQlQueryResources resourceType) {
+    return chunkedIds.stream()
+        .map(chunk -> new ResourceIdsGraphQlRequest(new HashSet<>(chunk), resourceType))
+        .collect(toList());
+  }
+
+  @Nonnull
+  protected Set<String> getNonCachedReferenceIds(@Nonnull final Set<String> referenceIds) {
+    return referenceIds.stream()
+        .filter(id -> !referenceIdToKeyCache.containsKey(id))
+        .collect(toSet());
+  }
+
+  protected void cacheResourceReferenceKeys(final Set<ResourceKeyId> results) {
+    Optional.ofNullable(results)
+        .orElseGet(Collections::emptySet)
+        .forEach(
+            resourceKeyId -> {
+              final String key = resourceKeyId.getKey();
```

Here as you were doing before, you could keep the placeholder as a key if it could be fetched from CTP, then during reference resolution, it could use the same error handling, which is also failing fast to error callback before executing actions.

---

https://github.com/commercetools/commercetools-sync-java/pull/697/commits/4c3892041cb3afa8f9a6167c31647b69aebf75cf